# EP Chan Book Implementation

This is a replicate of the strategy shown in the book "Algorithmic Trading Winning Strategies and their Rationale" by Ernest P. Chan. To use this notebook, you need specific virtual environment **qt-env** and setup the ipython kernel as "qt".

Mean reversion is a widely observed phenomenon in finance. Even though individual asset price may not follow mean reversion, but a combination of assets, i.e., a portfolio, may show mean reversion. Therefore, we can constucture a portfolio with mean reversion feature, and then trade on it to gain profit.

To do so, several things need to be done. First of all, we need to test whether a price series is mean reversion. The test is named **ADF test**. The test statistics is $\lambda$/SE($\lambda$), where $\lambda$ is the regression coefficient of equation:

$\Delta y(t) = \lambda y(t-1) +\mu + \beta t+\alpha_1\Delta y(t-1)+\dots+\alpha_k\Delta y(t-k)+\epsilon_t$.

Import necessary modules and print out version information.

In [2]:
import sys
sys.path.append('/home/adam/.pyenv/ep-chan/lib/python3.5/site-packages')

In [3]:
#import tushare as ts
import quandl
import pandas as pd
import seaborn as sns

from bokeh.plotting import figure
from bokeh.io import output_notebook, show
from bokeh.models import ColumnDataSource

from datetime import datetime

output_notebook()
pd.show_versions()

Loading BokehJS ...


INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-17134-Microsoft
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: None
pip: 8.1.1
setuptools: 40.8.0
Cython: None
numpy: 1.16.2
scipy: 1.2.1
pyarrow: None
xarray: None
IPython: 7.4.0
sphinx: None
patsy: 0.5.1
dateutil: 2.8.0
pytz: 2018.9
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 3.0.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: 4.2.1
bs4: 4.7.1
html5lib: 0.999
sqlalchemy: 1.2.7
pymysql: None
psycopg2: 2.7.7 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None


/home/adam/.pyenv/ep-chan/lib/python3.5/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)


The first task is to get a workable dataset. We can do this using *Tushare*.

In [4]:
import os

#df = ts.get_hist_data('sh', start='2014-04-30', end='2018-04-30', ktype='M') # get shanghai A share main index
df = quandl.get("WIKI/KO", start_date="2016-01-01", end_date="2018-01-01", api_key=os.environ['QUANDL_API_KEY'])
df = df.reset_index() # reset index of the dataframe
df['Date'] = pd.to_datetime(df['Date']) # convert str to datetime
print(df)

          Date   Open     High      Low  Close      Volume  Ex-Dividend  \
0   2016-01-04  42.34  42.5000  41.9650  42.40  16084145.0          0.0   
1   2016-01-05  42.31  42.6250  42.1700  42.55  15769463.0          0.0   
2   2016-01-06  42.20  42.5100  42.0400  42.32  16245132.0          0.0   
3   2016-01-07  41.65  42.1100  41.4496  41.62  20561556.0          0.0   
4   2016-01-08  41.65  42.1200  41.4100  41.51  14645738.0          0.0   
5   2016-01-11  41.58  41.8599  41.2900  41.58  15807748.0          0.0   
6   2016-01-12  42.30  42.4500  41.8700  42.12  21814837.0          0.0   
7   2016-01-13  42.17  42.5700  41.7900  41.85  16854282.0          0.0   
8   2016-01-14  41.99  42.1700  41.6400  41.88  16027468.0          0.0   
9   2016-01-15  40.95  41.5900  40.9260  41.50  21153542.0          0.0   
10  2016-01-19  41.77  42.1200  41.6500  41.92  15462140.0          0.0   
11  2016-01-20  41.45  41.6200  40.7500  41.38  19923495.0          0.0   
12  2016-01-21  41.45  41

## Stationary Test

### ADF test

ADF test is included in *statsmodels* module.

In [5]:
from statsmodels.tsa.stattools import adfuller
from bokeh.models import ColumnDataSource

adf_results = adfuller(df['Close'])
print(adf_results)

data = ColumnDataSource(df) # create object for bokeh plot
p = figure(x_axis_type='datetime', plot_width=300, plot_height=300)
p.line('Date', 'Close', line_width=2, source=data)
show(p)

(-2.0618550806365867, 0.2601267813146223, 0, 501, {'10%': -2.569851849873108, '1%': -3.443470002903981, '5%': -2.867326248740681}, 307.3872263966549)


The ADF test statistics is actually $\lambda/\text{SE}(\lambda)$, which is always negative for mean reversion. Here, the ADF test statistics is -3.595, for 95% confidence, we can reject the null hypothesis, that is, the "sh" index is stationary, even though from the plot it does not seem to be stationary.

$\lambda$ indicates the half-life of mean reversion, as half-life$=-log(2)/\lambda$. It can be determined by running a regression fit with $y(t)-y(t-1)$ as the dependent variable and $y(t-1)$ as the independent variable. 

### Hurst Exponent

Hurst exponent $H$ indicates whether the time series is stationary ($H<0.5$), random ($H=0.5$), or trending ($H>0.5$). First, we need to construct a function to [calculate hurst exponent](https://stackoverflow.com/questions/39488806/hurst-exponent-in-python).

In [6]:
from numpy import log10, polyfit, var, subtract

def hurst_ernie_chan(p, lag_range=None):

    p_log = log10(p) # use log price

    variancetau = []
    tau = []
    
    # Create the range of lag values
    if lag_range == None:
        lags = [2]
    else:
        lags = range(2, lag_range) # lag_range < len(ts)

    for lag in lags: 

        #  Write the different lags into a vector to compute a set of tau or lags
        tau.append(lag)

        # call this pp or the price difference
        pp = subtract(p_log[lag:], p_log[:-lag])
        variancetau.append(var(pp))

    # we now have a set of tau or lags and a corresponding set of variances.
    #print tau
    #print variancetau

    # plot the log of those variance against the log of tau and get the slope
    m = polyfit(log10(tau),log10(variancetau),1)

    hurst = m[0] / 2

    return hurst

print("Hurst(sh): {}".format(hurst_ernie_chan(df['Close'], 5)))

Hurst(sh): 0.3909245522396867


### Variance Ratio

Variance ratio can be used to test whether a financial return series is a pure random walk or having some predictability.

In [7]:
from arch.unitroot import VarianceRatio

vr = VarianceRatio(df['Close'], 5)
print(vr.summary().as_text())

     Variance-Ratio Test Results     
Test Statistic                 -3.225
P-value                         0.001
Lags                                5
-------------------------------------

Computed with overlapping blocks (de-biased)


The possible explanation of above test is as following (from E.P. Chan's book): test statistics equals to 1 means rejection of the random walk and test statistics equals to 0 means it may be a random walk. The p-value gives the smallest probability that you can reject the null hypothesis, which is a random walk. So in this case, with 95% confidence, we can reject the null hypothesis.

### Cointegration

For a pair of assets, we can run a CADF test to determine the cointegration. Let's see how to determine cointegration of two stocks: AAPL and IBM.

Here, we show how to get global market information via *pandas_datareader*.

In [8]:
pd.core.common.is_list_like = pd.api.types.is_list_like # https://stackoverflow.com/questions/50394873/import-pandas-datareader-gives-importerror-cannot-import-name-is-list-like
import pandas_datareader as web
from datetime import datetime

start = datetime(2014, 4, 30)
end = datetime(2018, 4, 30)

apple = web.DataReader('AAPL','iex', start, end) # IEX has stock prices up to 5 years
ibm = web.DataReader('IBM', 'iex', start, end)

5y
5y


Now, let's store the historical price information into database so that we don't need to request from the server every time.

In [9]:
from sqlalchemy import create_engine, MetaData, TEXT, Integer, Float, DateTime, Table, Column

# apple_table.drop(engine)
# ibm_table.drop(engine)

engine = create_engine("postgresql://postgres:@localhost/stocks") # stocks is the name of db
meta = MetaData(bind=engine)

apple_reset = apple.reset_index() # in order to create ID in mysql, we need to reset current index, 
apple_reset.index.name = 'ID'     # and then rename the index
apple_table = Table('apple', meta,
                 Column('ID', Integer, autoincrement=False),
                 Column('date', DateTime, nullable=False),
                 Column('open', Float, nullable=True),
                 Column('high', Float, nullable=True),
                 Column('close', Float, nullable=True),
                 Column('low', Float, nullable=True),
                 Column('volume', Float, nullable=True),
                 extend_existing=True
                )

ibm_reset = ibm.reset_index()
ibm_reset.index.name = 'ID'
ibm_table = Table('ibm', meta,
                 Column('ID', Integer, autoincrement=False),
                 Column('date', DateTime, nullable=False),
                 Column('open', Float, nullable=True),
                 Column('high', Float, nullable=True),
                 Column('close', Float, nullable=True),
                 Column('low', Float, nullable=True),
                 Column('volume', Float, nullable=True),
                 extend_existing=True
                )

meta.create_all(engine)

apple_reset.to_sql('apple', engine, if_exists='append', index=True)
ibm_reset.to_sql('ibm', engine, if_exists='append', index=True)

Now we can load data directly from database.

In [10]:
from sqlalchemy import create_engine

engine = create_engine("postgresql://postgres:@localhost/stocks") # stocks is the name of db
apple = pd.read_sql_table('apple', engine)
ibm = pd.read_sql_table('ibm', engine)

Once we have the stock prices, we can run a cointegration test (CADF test).

In [11]:
import statsmodels.tsa.stattools as ts

merged = pd.merge(apple, ibm, on = 'date')
x1 = merged['close_x']
y1 = merged['close_y']

cointegration = ts.coint(x1, y1)
print(cointegration)

(-1.1233407772843005, 0.8768488720575568, array([-3.89915898, -3.3376462 , -3.04550231]))


We found the test statistics is -1.124, greater than the 90% threshold -3.05, therefore, cannot reject the null hypothesis. That is, AAPL and IBM are not cointegration, as expected.

## Mean Reversion Strategies

### Bollinger bands

Bollinger bands is a practical mean reversion trading strategy. We enter the trade when price deviates *entryZscore* standard deviations from the mean. And we exit the position when price mean-reverts to *exitZscore* standard deviations from the mean, where *exitZscore* $<$ *entryZscore*.

In this example, we set *entryZscore*=1 and *exitZscore*=0, which means we will exit when the price mean-reverts to the current mean. Mean and standard deviation is calcualted within the lookback period.

Note, rolling OLS is removed from *Pandas*, we can use [another implementation](https://stackoverflow.com/questions/37317727/deprecated-rolling-window-option-in-ols-from-pandas-to-statsmodels) by Brad Solomon. This package works for *Python3*, so we need to make some local modifications in order to work with *Python27*.

**In order to compare results the E.P. Chan's book, from this point onwards, we will use datasets provided by E.P. Chan, and develop the same strategy as described on his book.**

First step is to load .mat data given by E.P. Chan. The .mat data in this case is three dimensional, containing 1500 dates, 67 symbols, and 4 attributes. Therefore, loading it as dataframe requires some extra work.

In [12]:
import scipy.io
import numpy as np
mat = scipy.io.loadmat('../data/inputData_ETF.mat') # load .mat file

""" this part is an attempt to load .mat to dataframe, but failed
however, the operations are useful as a reference
data = pd.DataFrame(data.items()) # convert dict to dataframe
data.columns = ['a', 'b'] # rename columns
data['b'].update(data['b'].apply(lambda l: [item for sublist in l for item in sublist])) # flatten list of lists
data.set_index('a',inplace=True) # use column a as index
data.loc['syms', 'b'] = [item for sublist in data.loc['syms', 'b'] for item in sublist] # flatten row syms and col b
data = data.drop(['__header__', '__globals__', '__version__']) # drop unecessary rows
print data
"""
vol = pd.DataFrame(np.hstack((mat['tday'], mat['vol']))) # use `np.hstack` to make mini dataframe
cl = pd.DataFrame(np.hstack((mat['tday'], mat['cl'])))
lo = pd.DataFrame(np.hstack((mat['tday'], mat['lo'])))
hi = pd.DataFrame(np.hstack((mat['tday'], mat['hi'])))
data = pd.concat([vol, cl, lo, hi], keys=['vol', 'cl', 'lo', 'hi']) # use `pd.concat` to combine mini dataframes
syms = [item for sublist in np.array(mat['syms']) for items in sublist for item in items] # flatten list of lists of lists for symbol names
col_names = ['tday']+syms # prepare the col names for the final dataframe
data.columns = col_names # reset the column names
print(data.head())

             tday  BGU  BGZ  BNO  DRN  DRV  EDC  EDZ       EEM  EEV  ...  VXZ  \
vol 0  20060426.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  101502.0  NaN  ...  NaN   
    1  20060427.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  121329.0  NaN  ...  NaN   
    2  20060428.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN   82875.0  NaN  ...  NaN   
    3  20060501.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN   72075.0  NaN  ...  NaN   
    4  20060502.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN   83598.0  NaN  ...  NaN   

           XLB       XLE       XLF      XLI      XLK      XLP      XLV  \
vol 0  22136.0  294271.0   54232.0   8592.0  24503.0   4504.0  21923.0   
    1  83830.0  448795.0  176893.0  14073.0  26474.0   6010.0  44698.0   
    2  65802.0  190539.0  133976.0  28273.0  29100.0  12658.0  24957.0   
    3  21191.0  195958.0  136147.0   7080.0  38514.0   7386.0  10186.0   
    4  33522.0  190415.0   39776.0   3700.0   8762.0  10278.0   7913.0   

           XLY  XME  
vol 0   8179.0  NaN  
    1  21670.0  NaN  
  

In [13]:
# massage the data
temp = data['tday'].apply(lambda x: str(int(x))) # convert float to str
data['tday'].update(temp) # update dataset, now 'tday' is stored as str
temp2 = data['tday'].apply(lambda x: datetime.strptime(x, '%Y%m%d')) # convert str to datetime object
data['tday'] = temp2 # update dataset, now 'tday' is datetime object
print(data.head())

            tday  BGU  BGZ  BNO  DRN  DRV  EDC  EDZ       EEM  EEV  ...  VXZ  \
vol 0 2006-04-26  NaN  NaN  NaN  NaN  NaN  NaN  NaN  101502.0  NaN  ...  NaN   
    1 2006-04-27  NaN  NaN  NaN  NaN  NaN  NaN  NaN  121329.0  NaN  ...  NaN   
    2 2006-04-28  NaN  NaN  NaN  NaN  NaN  NaN  NaN   82875.0  NaN  ...  NaN   
    3 2006-05-01  NaN  NaN  NaN  NaN  NaN  NaN  NaN   72075.0  NaN  ...  NaN   
    4 2006-05-02  NaN  NaN  NaN  NaN  NaN  NaN  NaN   83598.0  NaN  ...  NaN   

           XLB       XLE       XLF      XLI      XLK      XLP      XLV  \
vol 0  22136.0  294271.0   54232.0   8592.0  24503.0   4504.0  21923.0   
    1  83830.0  448795.0  176893.0  14073.0  26474.0   6010.0  44698.0   
    2  65802.0  190539.0  133976.0  28273.0  29100.0  12658.0  24957.0   
    3  21191.0  195958.0  136147.0   7080.0  38514.0   7386.0  10186.0   
    4  33522.0  190415.0   39776.0   3700.0   8762.0  10278.0   7913.0   

           XLY  XME  
vol 0   8179.0  NaN  
    1  21670.0  NaN  
    2   

Before doing any strategy, let's plot the close prices of USO and GLD.

In [14]:
data_close_2 = data.xs('cl')[['tday', 'USO', 'GLD']] # create a new df just for the 2 etfs
print(data_close_2.head())

# create bokeh data object
close_for_plot = ColumnDataSource(data_close_2) # convert pd.DataFrame to bokeh.models.ColumnDataSource

# create a figure object
p = figure(x_axis_type="datetime", title="Closing Prices")
p.grid.grid_line_alpha=0.3
p.xaxis.axis_label = 'Date'
p.yaxis.axis_label = 'Close Price'

# add elements to the figure
p.line('tday', 'USO', color='#A6CEE3', source=close_for_plot, legend='USO price')
p.line('tday', 'GLD', color='#FB9A99', source=close_for_plot, legend='GLD price')
p.legend.location = "top_left"

show(p)

        tday    USO    GLD
0 2006-04-26  69.54  63.65
1 2006-04-27  68.70  62.96
2 2006-04-28  69.62  65.09
3 2006-05-01  71.50  65.16
4 2006-05-02  72.34  66.55


Once we get the dataset straight, we can go ahead develop the strategy. We first need to run a rolling OLS to determine the hedge ratio between the two assets. Then, according to bollinger band definition, we find the trading signal and update the dollar position of each asset class.

In [17]:
import statsmodels.api as sm

y_ticker = 'USO'
x_ticker = 'GLD'

temp_y = data_close_2[y_ticker]
temp_x = data_close_2[x_ticker]

lookback = 20
entryZscore = 1.0
exitZscore = 0.0

data_close_2_copy = data_close_2 # copy the original data, to be safe
data_close_2_copy['hedge_ratio'] = np.nan # create a nan column for future hedge_ratio storage

for t in range(lookback, len(temp_x)): # loop through dates, for every date, calculate hedge ratio using lookback period
    y = temp_y[t-lookback:t] 
    x = temp_x[t-lookback:t]
    x = sm.add_constant(x) # add constant for `statsmodels.api.OLS` to work
    beta = sm.OLS(y,x).fit().params[1] # return slope from `OLS`
    data_close_2_copy.loc[t, 'hedge_ratio'] = beta # store beta as hedge ratio

print(data_close_2_copy.shape)
print(data_close_2_copy.head(30))

/home/adam/.pyenv/ep-chan/lib/python3.5/site-packages/numpy/core/fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


(1500, 4)
         tday    USO    GLD  hedge_ratio
0  2006-04-26  69.54  63.65          NaN
1  2006-04-27  68.70  62.96          NaN
2  2006-04-28  69.62  65.09          NaN
3  2006-05-01  71.50  65.16          NaN
4  2006-05-02  72.34  66.55          NaN
5  2006-05-03  70.22  66.46          NaN
6  2006-05-04  68.32  67.48          NaN
7  2006-05-05  68.00  67.99          NaN
8  2006-05-08  67.88  67.56          NaN
9  2006-05-09  68.40  69.68          NaN
10 2006-05-10  69.77  70.38          NaN
11 2006-05-11  70.30  71.03          NaN
12 2006-05-12  69.11  71.12          NaN
13 2006-05-15  66.63  67.41          NaN
14 2006-05-16  66.70  68.61          NaN
15 2006-05-17  65.70  68.15          NaN
16 2006-05-18  66.74  67.46          NaN
17 2006-05-19  65.57  65.58          NaN
18 2006-05-22  66.28  65.30          NaN
19 2006-05-23  67.77  66.38          NaN
20 2006-05-24  66.10  64.06    -0.008949
21 2006-05-25  67.80  64.70     0.110609
22 2006-05-26  67.75  65.10     0.163881
23 200

In [20]:
from numpy import matlib as mb

# book page 71, use price spread `USO - hedgeRatio * GLD` as signal
cols = [x_ticker, y_ticker]
yport = np.ones(data_close_2_copy[cols].shape)
yport[:,0] = -data_close_2_copy['hedge_ratio']
yport = yport * data_close_2_copy[cols]
yport = yport['GLD'] + yport['USO'] 

moving_avg = yport.rolling(window=lookback).mean()
moving_std = yport.rolling(window=lookback).std()
Zscore = (yport - moving_avg) / moving_std

# trade signals, boolean
long_entry = Zscore < -entryZscore # buy when price is lower than -entryZscore
long_exit = Zscore >= -exitZscore # sell when price is higher than or equal to -exitZscore
short_entry = Zscore > entryZscore # short buy when price is higher than entryZscore
short_exit = Zscore <= exitZscore # short sell when price is lower than or equal to exitZscore

num_units_long = np.empty((len(yport), 1))
num_units_long = pd.DataFrame(np.where(long_entry, 1, 0))
# num_units_long = pd.DataFrame(np.where(long_exit, 0, np.nan))
num_units_short = np.empty((len(yport), 1))
num_units_short = pd.DataFrame(np.where(short_entry, -1, 0))
# num_units_short = pd.DataFrame(np.where(short_exit, 0, np.nan))
num_units = num_units_long + num_units_short

# create dollar position for each asset
temp1 = pd.DataFrame(mb.repmat(num_units, 1, 2))
temp2 = np.ones(data_close_2_copy[cols].shape)
temp2[:, 0] = -data_close_2_copy['hedge_ratio']

position = np.multiply(np.multiply(temp1, temp2), data_close_2_copy[cols])
position.columns = [x_ticker, y_ticker]
print(position.tail(10))

            GLD    USO
1490  -0.000000   0.00
1491  -0.000000   0.00
1492  -0.000000   0.00
1493  -0.000000   0.00
1494  -0.000000   0.00
1495  -0.000000   0.00
1496  -0.000000   0.00
1497 -17.044724  38.84
1498 -25.320990  39.26
1499 -28.921578  38.97


Next, let's compute the P&L and other performance measures for this strategy.

In [1]:
# daily pnl in dollar
pnl = np.sum(np.multiply(position[:-1], np.divide(np.diff(data_close_2_copy[cols], axis=0), data_close_2_copy[cols][:-1])), axis=1)

# gross market value
mkt_value = pd.DataFrame.sum(abs(position[:-1]), axis=1)
print(mkt_value)

# return is pnl divided by gross market value
ret = pnl / mkt_value
ret = ret.fillna(method='pad')

# compute Sharpe
sharpe = (np.sqrt(252) * np.mean(ret)) / np.std(ret) 
APR = np.prod(1+ret) ** (252.0 / len(ret)) - 1
print('Price spread Sharpe: {:.4}'.format(sharpe))
print('Price spread APR: {:.4%}'.format(APR))

NameError: name 'np' is not defined

Now, let's plot the cumulative returns.

In [22]:
acum_ret = ret.cumsum() # get cumulative return
acum_ret = acum_ret.fillna(method='pad') # fill with pad
acum_ret = acum_ret.fillna(0) # fill the rest NAN with 0

acum_ret_date = pd.concat([data_close_2_copy['tday'][1:].reset_index(drop=True), acum_ret], axis=1) # concat date and acum_ret
acum_ret_date.columns = ['tday', 'acum_ret'] # reset column names

acum_ret_for_plot = ColumnDataSource(acum_ret_date) # convert pd.DataFrame to bokeh.models.ColumnDataSource

p1 = figure(x_axis_type="datetime", title="Cumulative Returns")
p1.grid.grid_line_alpha=0.3
p1.xaxis.axis_label = 'Date'
p1.yaxis.axis_label = 'Cumulative Returns'

p1.line('tday', 'acum_ret', color='#A6CEE3', legend='acum return', source=acum_ret_for_plot)
p1.legend.location = "top_left"

show(p1)

### Kalman filter

Intead of OLS, we can use Kalman filter to find the optimal hedge ratio and moving average. Kalman filter is a linear algorithm to update the expected value of hidden variable based on the latest value of an observable variable. In our case, 

$y(t) = x(t)\beta(t)+\epsilon(t)$,

the asset prices $y$ is the observable variable, whereas $\beta$ is the hidden variable. Note, $\beta$ contains two parts, the slope and intercept, where slope represents the hedge ratio and intercept represents the moving average of the spread.

Kalman filter has been implemented in *pykalman* package.

Read more about Kalman filter in trading: [QuantStart Dynamic Hedge Ratio](https://www.quantstart.com/articles/Dynamic-Hedge-Ratio-Between-ETF-Pairs-Using-the-Kalman-Filter) and [Quantopian Kalman filter lecture](https://www.youtube.com/watch?v=RxIdLu18SsE).

Since data has been loaded and massaged in the last section, we can directly use the dataset.

In [24]:
from pykalman import KalmanFilter

data_kf = data[['tday', 'EWA', 'EWC']]
data_kf_cl = data_kf.xs('cl')
etfs = ['EWA', 'EWC']

def calc_slope_intercept_kalman(etfs, prices):
    """
    Taken directly from QuantStart Dynamic Hedge Ratio website
    Utilise the Kalman Filter from the pyKalman package
    to calculate the slope and intercept of the regressed
    ETF prices.
    """
    delta = 1e-5 # delta is a param to control the transition covariance matrix; delta=0 means linear reg
    trans_cov = delta / (1 - delta) * np.eye(2)
    obs_mat = np.vstack(
        [prices[etfs[0]], np.ones(prices[etfs[0]].shape)]
    ).T[:, np.newaxis]

    kf = KalmanFilter(
        n_dim_obs=1,
        n_dim_state=2,
        initial_state_mean=np.zeros(2),
        initial_state_covariance=np.ones((2, 2)),
        transition_matrices=np.eye(2),
        observation_matrices=obs_mat,
        observation_covariance=1.0,
        transition_covariance=trans_cov
    )

    state_means, state_covs = kf.filter(prices[etfs[1]].values)
    return state_means, state_covs

state_means, state_covs = calc_slope_intercept_kalman(etfs, data_kf_cl)
print(state_means)

[[1.3375311  1.3375311 ]
 [1.33953995 1.33953476]
 [1.34020571 1.34019542]
 ...
 [1.15051616 1.59479389]
 [1.14905394 1.59467038]
 [1.14760939 1.59449835]]


Once we have state means ($\beta$), we can go ahead plot the optimal hedge ratio and moving average of the spread.

In [26]:
state_means_temp = pd.DataFrame(np.array(state_means))
state_means_df = pd.concat([data_kf_cl['tday'], state_means_temp], axis=1)
state_means_df.columns = ['tday', 'hedge ratio', 'moving avg']

kf_for_plot = ColumnDataSource(state_means_df) # convert pd.DataFrame to bokeh.models.ColumnDataSource

p1 = figure(x_axis_type="datetime", title="Slope/Hedge Ratio")
p1.grid.grid_line_alpha=0.3
p1.xaxis.axis_label = 'Date'
p1.yaxis.axis_label = 'Hedge Ratio'

p1.line('tday', 'hedge ratio', color='#A6CEE3', legend='slope/hedge ratio', source=kf_for_plot)
p1.line('tday', 'moving avg', color='#FB9A99', legend='intercept/moving avg', source=kf_for_plot)
p1.legend.location = "top_left"

show(p1)

### Cross-sectional mean reversion

Cross-sectional mean reversion relies on the reversion of short-term relative returns of a basket of assets (usually stocks, not futures or currencies). The serial anti-correlation of the relative returns can generate profits. The relative return is the return of individual stock minus the average return of the stocks in a particular universe. 

We determine the weights of every stock using

$w_i = -\frac{r_i -<r_j>}{\sum_{k}|r_k-<r_j>|}$

where $r_i$ is the daily return of $i$th stock, and $<r_j>$ is the average daily return of all stocks in the index. If a stock has very positive return, we will short lots of it, and if it has a very negative return, we may buy it.

For this section, we need a new set of data.

In [28]:
import scipy.io
import numpy as np
from datetime import datetime

mat = scipy.io.loadmat('../data/inputDataOHLCDaily_20120504.mat') # load .mat file

syms = [item for sublist in np.array(mat['syms']) for items in sublist for item in items] # flatten list of lists of lists for symbol names
tday = pd.DataFrame(mat['tday'])
tday.columns = syms
cl = pd.DataFrame(mat['cl'])
cl.columns = syms
# op = pd.DataFrame(mat['op'])
# op.columns = syms
print(cl.head())

      AUD     AXJ     BO     BRE      BZ     CAD      CJ      CL       C  \
0  0.8097  5606.0  68.50  0.6149  149.00  0.9968  2918.0  175.48  674.25   
1  0.8035  5708.0  69.33  0.6140  151.78  0.9959  2910.0  177.74  677.25   
2  0.8091  5723.0  71.55  0.6144  156.64  0.9933  2887.0  181.93  694.75   
3  0.8064  5734.0  70.13  0.6114  155.46  0.9870  2881.0  179.57  683.25   
4  0.8101  5671.0  71.86  0.6174  155.51  0.9915  2884.0  180.95  687.25   

     EMD  ...     TF      TT        TU        US     VX       W      YO  \
0  888.6  ...  731.9  0.4732  100.6016  101.4844  43.29  1204.5  0.0700   
1  874.7  ...  707.5  0.4628  100.3516  100.3281  44.35  1197.5  0.0677   
2  874.7  ...  703.9  0.4649  100.6328  101.3750  45.67  1191.5  0.0674   
3  863.6  ...  697.8  0.4550  100.5078  100.1563  45.25  1158.5  0.0657   
4  871.2  ...  687.8  0.4472  100.3125   99.2813  45.76  1166.0  0.0615   

         ZB        ZF      ZN  
0  106.6094  102.3359  96.266  
1  105.3594  101.7109  95.37

But *tday* dataframe is in the type of double, rather than datetime. In order to convert the entire dataframe to datetime, let's write a function.

In [32]:
def convert_df_to_datetime(df, datetime_format_str='%Y%m%d'):
    """Convert `double` to `datetime`.
    
    @param df: pd.DataFrame
    @param datetime_format_str: str, represents the format of datetime
    @return: pd.DataFrame
    """

    ncol = len(df.columns)
    for i in range(ncol):
        col_name = df.columns[i]
#         index_nan = df[col_name].index[df[col_name].apply(np.isnan)]
#         print index_nan
        df[col_name] = df[col_name].fillna(10000101.0) # for nan, use 1000-01-01 as a place holder
        temp = df[col_name].apply(lambda x: str(int(x))) # convert float to str
        df[col_name].update(temp) # update dataset, now `df` is stored as str
        temp2 = df[col_name].apply(lambda x: datetime.strptime(x, datetime_format_str)) # convert str to datetime object
        df[col_name] = temp2 # update dataset, now `df` is datetime object
    
    return df

In [30]:
def convert_series_to_datetime(series, datetime_format_str='%Y%m%d'):
    """Convert `double` to `datetime`.
    
    @param series: pd.Series
    @param datetime_format_str: str, represents the format of datetime
    @return: pd.Series
    """

    series = series.fillna(10000101.0) # for nan, use 1000-01-01 as a place holder
    temp = series.apply(lambda x: str(int(x))) # convert float to str
    series.update(temp) # update dataset, now `df` is stored as str
    temp2 = series.apply(lambda x: datetime.strptime(x, datetime_format_str)) # convert str to datetime object
    series = temp2 # update dataset, now `df` is datetime object
    
    return series

The dataset can be converted to several dataframes. One of them is "tday" dataframe, which includes different trading dates for every single symbol. However, in order to build this strategy, we need the entire basket of symbols. So we need to set a start and end date that all stock symobls have, and get the "cl" data between these dates.

In [33]:
tday_datetime = convert_df_to_datetime(tday)

start_day = '20080522'
end_day = '20120430'
start_mask = tday_datetime['AUD'] == start_day # use boolean mask to find index of a value within dataframe
end_mask = tday_datetime['AUD'] == end_day
idx_start = tday_datetime['AUD'][start_mask].index[0]
idx_end =  tday_datetime['AUD'][end_mask].index[0]

cl = cl.loc[idx_start:idx_end] # select range of index from "cl" dataframe
print(cl.head())
print(cl.tail())

      AUD     AXJ     BO     BRE      BZ     CAD      CJ      CL       C  \
1  0.8035  5708.0  69.33  0.6140  151.78  0.9959  2910.0  177.74  677.25   
2  0.8091  5723.0  71.55  0.6144  156.64  0.9933  2887.0  181.93  694.75   
3  0.8064  5734.0  70.13  0.6114  155.46  0.9870  2881.0  179.57  683.25   
4  0.8101  5671.0  71.86  0.6174  155.51  0.9915  2884.0  180.95  687.25   
5  0.8037  5594.0  70.50  0.6214  152.25  0.9928  2907.0  177.61  685.50   

     EMD  ...     TF      TT        TU        US     VX       W      YO  \
1  874.7  ...  707.5  0.4628  100.3516  100.3281  44.35  1197.5  0.0677   
2  874.7  ...  703.9  0.4649  100.6328  101.3750  45.67  1191.5  0.0674   
3  863.6  ...  697.8  0.4550  100.5078  100.1563  45.25  1158.5  0.0657   
4  871.2  ...  687.8  0.4472  100.3125   99.2813  45.76  1166.0  0.0615   
5  880.0  ...  690.4  0.4176  100.2656   98.4531  45.02  1172.5  0.0621   

         ZB        ZF      ZN  
1  105.3594  101.7109  95.375  
2  106.1094  102.2031  96.03

Now, we have massaged data, we can go ahead calculate return and weights.

In [34]:
cl_lag = cl.shift(-1)
daily_return = np.divide(np.subtract(cl, cl_lag), cl_lag)
mkt_return = np.mean(daily_return, axis=1) # market return 

mkt_return_list = np.tile(mkt_return, (daily_return.shape[1], 1)) # like repmat in *Matlab*, creates copies of an item
mkt_return_df = pd.DataFrame(mkt_return_list)
mkt_return_df = mkt_return_df.T
mkt_return_df.columns = syms
numerator = -(np.subtract(daily_return, mkt_return_df))

denominator_temp = np.sum(np.abs(numerator), axis=1)
denominator_list = np.tile(denominator_temp, (daily_return.shape[1], 1))
denominator = pd.DataFrame(denominator_list)
denominator = denominator.T
denominator.columns = syms

weights = np.divide(numerator, denominator)

final_return = np.sum(np.multiply(weights.shift(1), daily_return), axis=1) # shift weights downwards so that final return is calculated as yesterday's weights times today's return
acum_return = np.cumprod(1 + final_return) - 1

Once we have the cumulative return calculated, we can evaluate the performance.

In [35]:
# compute performance
sharpe = (np.sqrt(252) * np.mean(final_return)) / np.std(final_return) 
APR = np.prod(1+final_return) ** (252.0 / len(final_return)) - 1
print('Price spread Sharpe: {:.4}'.format(sharpe)) 
print('Price spread APR: {:.4%}'.format(APR))

acum_return_date = pd.concat([tday_datetime['AUD'][idx_start:idx_end].reset_index(drop=True), acum_return[:-2]], axis=1) # concat date and acum_ret
acum_return_date.columns = ['tday', 'acum_return'] # reset column names
acum_return_for_plot = ColumnDataSource(acum_return_date) # convert pd.DataFrame to bokeh.models.ColumnDataSource

p1 = figure(x_axis_type="datetime", title="Cumulative Returns")
p1.grid.grid_line_alpha=0.3
p1.xaxis.axis_label = 'Date'
p1.yaxis.axis_label = 'Cumulative Returns'

p1.line('tday', 'acum_return', color='#A6CEE3', legend='acum ret', source=acum_return_for_plot)
p1.legend.location = "top_left"

show(p1)

Price spread Sharpe: 0.9214
Price spread APR: 12.7533%


### Trading calendar spread

Futures usually do not show mean reversion. However, for a specific time period, currency pairs exist mean reversion. Therefore, we can use similar mean reversion strategy described above to trade currency cross-rates.

To do so, the first task is to identify a currency pair that is mean reverting. This has been discussed in Chen's book. Here we will try to recreate a mean reversion calendar spread strategy.

In [37]:
import scipy.io
import numpy as np
from datetime import datetime

mat = scipy.io.loadmat('../data/inputDataDaily_CL_20120813.mat') # load .mat file
contracts = [item for sublist in np.array(mat['contracts']) for items in sublist for item in items] # flatten list of lists of lists for symbol names
tday = pd.DataFrame(mat['tday'])
tday.columns = ['tday']
cl = pd.DataFrame(mat['cl'])
cl.columns = contracts

# get spot price, clean contracts list and cl
temp = pd.concat([cl, tday], axis=1)
spot = temp[['tday','0000$']]
contracts = contracts.remove('0000$')
cl = cl.drop('0000$', axis=1)
print(cl.head())

   2007F  2007G  2007H  2007J  2007K  2007M  2007N  2007Q  2007U  2007V  ...  \
0    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  ...   
1    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  ...   
2    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  ...   
3    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  ...   
4    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  ...   

   2013Q  2013U  2013V  2013X  2013Z  2014F  2014G  2014H  2014J  2014K  
0    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
1    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
2    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
3    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
4    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  

[5 rows x 89 columns]


In this dataset, the first column is the spot price. The last column is the trading date. The rest is different contracts. We first extract the spot price, pick a date and fit the prices of futures of five nearest maturities to their time-to-maturity $T$. In this way, we get $\gamma$.

In [40]:
import statsmodels.api as sm

spot['gamma'] = np.nan
for t in range(spot.shape[0]):
    ft = cl.iloc[t]
    idx = ft.reset_index().index
    idx_list = idx.tolist()
    idx_list_no_na = [idx_list[i] for i in range(len(ft)) if not ft.isna()[i]] # get rid of NAN ones
    idx_no_na = pd.Series(idx_list_no_na)
    idx_diff = idx_no_na.shift(-1) - idx_no_na
    if len(idx_no_na) >= 5 and any(idx_diff[0:3]==1):
        ft_new = list(ft[idx_no_na[0:4]]) # use list instead of pd.Series so that sm.OLS can handle it
        tt = range(len(ft_new))
        tt = sm.add_constant(tt) # add constant for `statsmodels.api.OLS` to work
        beta = sm.OLS(ft_new,tt).fit().params[1] # return slope from `OLS`
        spot.loc[t, 'gamma'] = -12 * beta
print(spot['gamma'].tail())

/home/adam/.pyenv/ep-chan/lib/python3.5/site-packages/ipykernel-4.8.2-py3.5.egg/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


KeyboardInterrupt: 

Known $\gamma$, we can find the half-life of it. 

In [ ]:
# TODO

To apply mean reversion strategy, we need to find Zscore, with lookback set equal to the half-life.

In [ ]:
# TODO

A pair of contract is selected based on a set of criteria:

* First, the holding period for a pair of contracts is 3 months (61 trading days).
* Second, we roll forward to the next pair of contracts 10 days before the current near contract’s expiration.
* Third, the expiration dates of the near and far contracts are 1 year apart.

Assume we hold a long position in the far contract, and a short contract in the near one initially, we can find the true position according to our mean reversion strategy.

In [ ]:
# TODO

We can calculate the performance and plot the cumulative return.

In [ ]:
# TODO

## Momentum Strategies

### Cross-sectional momentum

The section aims to create a momentum strategy that buys and holds stocks within the top decile of 12-month lagged returns for a month, and vice versa for the bottom decile. In this way, we trade on a scross-sectional momentum of stocks.

As usual, the first task is to get the data in a workable format.

In [41]:
import scipy.io
import numpy as np
from datetime import datetime

mat = scipy.io.loadmat('../data/inputDataOHLCDaily_stocks_20120424.mat') # load .mat file
syms = [item for sublist in np.array(mat['stocks']) for items in sublist for item in items] # flatten list of lists of lists for symbol names
tday = pd.DataFrame(mat['tday'])
tday_datetime = convert_df_to_datetime(tday)
tday_datetime.columns = ['tday']
cl = pd.DataFrame(mat['cl'])
cl.columns = syms
# op = pd.DataFrame(mat['op'])
# op.columns = syms
print(cl.tail())
print(tday_datetime.tail())

          A    AA    AAPL    ABC    ABT    ACE    ACN   ADBE    ADI    ADM  \
1495  42.32  9.95  608.34  37.45  60.46  72.35  63.56  33.39  38.39  30.93   
1496  41.92  9.76  587.44  37.48  59.51  74.84  63.05  32.90  38.28  30.68   
1497  42.10  9.70  572.98  37.75  59.88  75.30  63.40  33.06  37.81  30.71   
1498  41.27  9.63  571.70  37.49  59.87  74.50  63.03  32.62  37.22  30.42   
1499  39.80  9.66  560.28  37.50  60.73  75.24  62.82  32.40  37.46  30.94   

      ...     XL   XLNX    XOM   XRAY   XRX    XYL   YHOO    YUM   ZION    ZMH  
1495  ...  20.89  35.37  85.75  40.31  7.96  27.65  15.49  72.94  21.17  64.29  
1496  ...  21.37  35.04  85.28  39.73  7.91  27.10  15.40  71.41  20.93  63.06  
1497  ...  21.38  34.40  85.30  39.73  7.87  27.36  15.60  73.93  20.54  63.08  
1498  ...  21.03  33.92  85.69  39.27  7.88  26.89  15.33  73.78  20.81  62.13  
1499  ...  21.26  33.53  86.31  39.98  7.92  27.10  15.43  72.24  20.06  62.39  

[5 rows x 497 columns]
           tday
1495 

Given the dataset, we can go ahead create the long and short signals. We calculate the return for each day and sort the returns, and create a long and a short array respectively.

In [43]:
lookback = 252
holddays = 25
topN = 50

ret=(cl - cl.shift(lookback)) / cl.shift(lookback)

# initialize signal matrix
longs = pd.DataFrame().reindex_like(cl)
shorts = pd.DataFrame().reindex_like(cl)

for t in range(lookback+1, len(tday)):
    temp = ret.iloc[t]
    temp_sort = temp.sort_values()
    idx = temp_sort.index
    idx_list = idx.tolist()
    idx_list_no_na = [idx_list[i] for i in range(len(temp_sort)) if temp_sort.isna()[i] == False] # get rid of NAN ones
    longs.at[t, idx_list_no_na[-topN:]] = True
    shorts.at[t, idx_list_no_na[0:topN]] = True

Then, we create a zero *positions* matrix. We lag the long and short signal according to our holding period. Update the *positions* matrix for every holding day.

In [45]:
positions = pd.DataFrame().reindex_like(longs)
positions = positions.fillna(0)

for h in range(holddays):
    long_lag = longs.shift(h)
    long_lag = long_lag.fillna(0)
    long_lag = long_lag.astype('int')
    
    short_lag = shorts.shift(h)
    short_lag = short_lag.fillna(0)
    short_lag = short_lag.astype('int')
    
    positions = positions + long_lag.where(long_lag == 0, 1)
    positions = positions + short_lag.where(short_lag == 0 , -1)

Once the *positions* matrix is obtained, we can trade accordingly. The daily and cumulative return can be calculated. Sharpe ratio and APR can be obtained.

In [46]:
dailyret=np.sum(positions.shift(1) * (cl - cl.shift(1)) / cl.shift(1), axis=1) / (2 * topN) / holddays
dailyret.fillna(0)

# select specific start and end date
start_day = '20070515'
end_day = '20071231'
start_mask = tday_datetime['tday'] == start_day
end_mask = tday_datetime['tday'] == end_day
idx_start = tday_datetime['tday'][start_mask].index[0]
idx_end =  tday_datetime['tday'][end_mask].index[0]

# calculate cumulative return
cumret=np.cumprod(1 + dailyret[idx_start:idx_end]) - 1;

sharpe = (np.sqrt(252) * np.mean(dailyret[idx_start:idx_end])) / np.std(dailyret[idx_start:idx_end]) 
APR = np.prod(1+dailyret[idx_start:idx_end]) ** (252.0 / len(dailyret[idx_start:idx_end])) - 1
print('Price spread Sharpe: {:.4}'.format(sharpe))
print('Price spread APR: {:.4%}'.format(APR))

Price spread Sharpe: 4.202
Price spread APR: 38.6203%


Finally, we go ahead plot the cumulative return.

In [47]:
acum_return_date = pd.concat([tday_datetime['tday'][idx_start:idx_end].reset_index(drop=True), cumret.reset_index(drop=True)], axis=1) # reset_index so that two df can be concated correctly
acum_return_date.columns = ['tday', 'acum_return'] # reset column names
acum_return_for_plot = ColumnDataSource(acum_return_date) # convert pd.DataFrame to bokeh.models.ColumnDataSource

p1 = figure(x_axis_type="datetime", title="Cumulative Returns")
p1.grid.grid_line_alpha=0.3
p1.xaxis.axis_label = 'Date'
p1.yaxis.axis_label = 'Cumulative Returns'

p1.line('tday', 'acum_return', color='#A6CEE3', legend='acum ret', source=acum_return_for_plot)
p1.legend.location = "top_left"

show(p1)

## Conclusion

So far, we have implemented most important contents in E.P. Chan's book. For every part of code, I added some necessary explanations to show the thinking process. This serves as a starting point of quantitative trading. One should read more books and articles to find out his/her own strategy. Morevoer, to gain a solid understanding, try to implement a backtesting environment yourself before going to some well known places, such as Quantopian and WorldQuant.

Hope everyone find their way to success!